etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/transformation/TransformationRule.java

```diff
@@ -0,0 +1,83 @@
+
+import gov.hhs.cdc.trustedintermediary.etor.ruleengine.FhirResource;
+import gov.hhs.cdc.trustedintermediary.etor.ruleengine.Rule;
+import gov.hhs.cdc.trustedintermediary.etor.ruleengine.RuleExecutionException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * The TransformationRule class extends the {@link
+ * gov.hhs.cdc.trustedintermediary.etor.ruleengine.Rule Rule} class and represents a transformation
+ * rule. It implements the {@link
+ * gov.hhs.cdc.trustedintermediary.etor.ruleengine.Rule#runRule(FhirResource) runRule} method to
+ * apply a transformation to the FHIR resource.
+ */
+public class TransformationRule extends Rule<TransformationRuleMethod> {
+
+    //    private static final Map<String, Class<?>> classCache = new ConcurrentHashMap<>();
+    private static final Map<String, CustomFhirTransformation> transformationInstanceCache =
+            new ConcurrentHashMap<>();
+
+    /**
+     * Do not delete this constructor! It is used for JSON deserialization when loading rules from a
+     * file.
+     */
+    public TransformationRule() {}
+
+    public TransformationRule(
+            String ruleName,
+            String ruleDescription,
+            String ruleMessage,
+            List<String> ruleConditions,
+            List<TransformationRuleMethod> ruleActions) {
+        super(ruleName, ruleDescription, ruleMessage, ruleConditions, ruleActions);
+    }
+
+    @Override
+    public void runRule(FhirResource<?> resource) throws RuleExecutionException {
+        for (TransformationRuleMethod transformation : this.getRules()) {
+            applyTransformation(transformation, resource);
+        }
+    }
+
+    private void applyTransformation(
+            TransformationRuleMethod transformation, FhirResource<?> resource)
+            throws RuleExecutionException {
+        String name = transformation.name();
+        Map<String, String> args = transformation.args();
+        logger.logInfo("Applying transformation: ", name);
+
+        CustomFhirTransformation transformationInstance = getTransformationInstance(name);
+        transformationInstance.transform(resource, args);
+    }
+
+    static CustomFhirTransformation getTransformationInstance(String name) throws RuntimeException {
+        return transformationInstanceCache.computeIfAbsent(
+                name, TransformationRule::createTransformationInstance);
+    }
+
+    private static CustomFhirTransformation createTransformationInstance(
+            String transformationName) {
+        String fullClassName = getFullClassName(transformationName);
+        try {
+            Class<?> clazz = Class.forName(fullClassName);
+            return (CustomFhirTransformation) clazz.getDeclaredConstructor().newInstance();
+        } catch (ClassNotFoundException
+                | InstantiationException
```

Comment:
Would it be better to catch an exception here instead of trying to catch the many different errors that could come through?

Reply:
We'd loose more specific information about the cause of the error if we just catch one Exception

Reply:
I wonder if we would benefit from having a separate log entry for each kind of exception. It might be robust but it will be more detailed.

Reply:
It would still print the specific exception type

Reply:
In that case, I'll collapse it into one exception. I don't see the benefit of logging each exception separately unless we're adding information that is not already in the exception itself

Reply:
Done